Keyboard and Screenreader a11y improvements

bgareth opened the pull request:
This PR is about implementing keyboard and screenreader a11y improvements.

Notable changes:
- Added screen reader only elements
  - Shortcut list element (before the toolbar / workbook in DOM)
  - Alert element to read out selection coordinates clearly when navigating the grid
  - Alert element to read when sheet focus lock is toggled
- Added ability to toggle sheet focus lock as some screen readers, such as NVDA, conflict with the sheet keyboard controls and cannot access the toolbar and other elements easily. Most sheet specific shortcuts and keyboard navigation is ignored when toggle off, allowing for regular keyboard navigation to take place
- Added more aria-label depth to toolbar items and other intractable elements to improve reading
- Added shortcuts for keyboard support for "drag and drop auto-fill"
- Removed default Canvas "image" aria reading as it was unhelpful
- Added strings and translations to account for the above
- On sheet mount focus on cell A1 allows for keyboard initial navigation to the sheet, as before the navigation controls were nullified without any selection in focus.
- eslint fixes and update to delete test in order to account for changes and runner errors
- Addressed some A11y DOM standards flagged by Axe Dev tools

Notes:
- This is not intended to change the experience for non screen reader and keyboard users, but only to improve the clarity and available functionality for screen reader and keyboard users.
- New shortcuts were chosen to avoid popular Mac, Windows and Chrome actions as best as possible while keeping in Google Sheet and standard practices. 
- Tested using NVDA, MacVO and Chrome screen reader.



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207501536719819

sanchit3008 commented:
thanks for this @bgareth , please let me know when this is ready for review!

bgareth commented:
> thanks for this @bgareth , please let me know when this is ready for review!

Thank you so much, @sanchit3008. This is ready for review. Please let me know if you have any questions or requests. 
I will also get another set of eyes from my team.

Separate note: once we are happy with the state of a11y for this project, we intend to add significant translation support as discussed last year.

sanchit3008 commented:
thanks @bgareth , i will start reviewing this tomorrow

@Corbe30 you can review this as well

bgareth commented:
I noticed another weakness in keyboard functionality around accessing the cell range context menu. I will get working on a commit, but if this review is complete by the time I am done, then I will separate it out to a new PR.

bgareth commented:
> You can include this in the pr itself, however it's going to take me some time to review due to personal commitments

Thanks, @sanchit3008. No rush.

Corbe30 commented:
Thanks @bgareth. I'll review the PR today ✨

bgareth commented:
> <img alt="image" width="488" src="https://private-user-images.githubusercontent.com/53386582/413877359-eb0d7f63-bd37-4868-9700-01a9fe310f4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4MDExODQsIm5iZiI6MTczOTgwMDg4NCwicGF0aCI6Ii81MzM4NjU4Mi80MTM4NzczNTktZWIwZDdmNjMtYmQzNy00ODY4LTk3MDAtMDFhOWZlMzEwZjRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE3VDE0MDEyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY0NWRhMWY3MjIwYzI2NmU4NThmY2JiZjc1MGQyZDFlYzlhY2YzNjliMTljOTg5NjYwYzY2ZGUwZDNkMDFlM2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cbBuGBoINU9_kJrYztqkxxitgqDDHyme-YtCqxEzY6I">
> In this PR, adding new sheets pushes the new sheet tabs vertically instead of horizontally. Can you please fix it @bgareth

Thanks for catching this, @Corbe30. I have pushed a commit that addresses this.

Corbe30 commented on packages/react/src/components/SheetOverlay/index.tsx:
```diff
@@ -825,7 +895,19 @@ const SheetOverlay: React.FC = () => {
         </div>
       </div>
-    </div>
+      <div id="sr-selection" className="sr-only" role="alert">
```
I'm not sure how to trigger this div, can you help me with this @bgareth

Corbe30 replied:
Got it, I didn't know about these tools, thanks!

Corbe30 commented:
Thanks for fixing the sheet tabs bug @bgareth! I've left some new comments, can you please fix those too

bgareth commented on packages/core/src/events/keyboard.ts:
```diff
@@ -682,6 +768,37 @@ export function handleGlobalKeyDown(
   // }
 
+  // Toggle focus with Ctrl + Shift + F (independent of sheet focus state)
+  if (e.ctrlKey && e.shiftKey && kstr === "F") {
```
I'm happy to explore improvements here if needed. However, I left this action in a different scope so it can happen despite `ctx.sheetFocused` and the `if (!ctx.sheetFocused) {return}` could catch everything else.

bgareth commented on packages/react/src/components/SheetOverlay/index.tsx:
```diff
@@ -321,6 +326,12 @@ const SheetOverlay: React.FC = () => {
     );
   }, [context, rightclick.rowOverLimit, setContext, showAlert]);
 
+  useEffect(() => {
```
Strangely, I was unable to replicate this issue. However I can see how the code would suggest the behaviour. I have applied a fix, but my experience does not change. Please verify.

Corbe30 replied:
Verified 👍

bgareth commented:
> Thanks for fixing the sheet tabs bug @bgareth! I've left some new comments, can you please fix those too

Thanks again, @Corbe30. I have addressed the feedback. I have left responses on a couple individual comments.

bgareth commented:
Hi there @sanchit3008. I hope you are well.

I wanted to provide a quick update regarding our use of your library and the contributions we had planned. As we mentioned previously, we were aiming to contribute a significant translation update. We're still very grateful for your support and the work you've put into the project, and I didn’t want to leave you wondering about the status on our end.

The project we were working on was part of an initiative under the U.S. Department of Defense, which has unfortunately been significantly impacted by recent political developments. As a result, our team has had to deprioritize the project for now. We're currently exploring ways to restructure it under a different umbrella, but that process is still in flux.

sanchit3008 commented:
hi @bgareth, that a big letdown! though to be honest i have not been able to devote time to this project lately as well. im hoping that you are personally keeping well during this period of confusion.
